src/v/ssx/metrics.cc

```diff
@@ -0,0 +1,141 @@
+
+ssmi::metric_function get_op(agg_op op, registered_instances refs) {
+    switch (op) {
+    case agg_op::sum:
+        return [refs{std::move(refs)}]() {
+            auto type = refs.front()->get_function()().type();
+            return absl::c_accumulate(
+              refs, ssmi::metric_value(0, type), [](auto lhs, auto rhs) {
+                  return lhs + rhs->get_function()();
+              });
+        };
+    case agg_op::count:
+        return [s = static_cast<double>(refs.size())]() {
+            return ssmi::metric_value(s, ssmi::data_type::ABSOLUTE);
+        };
+    case agg_op::min:
+        return [refs{std::move(refs)}]() {
+            auto it = absl::c_min_element(refs, [](auto lhs, auto rhs) {
+                return lhs->get_function()().d() < rhs->get_function()().d();
+            });
+            return (*it)->get_function()();
+        };
+    case agg_op::max:
+        return [refs{std::move(refs)}]() {
+            auto it = absl::c_max_element(refs, [](auto lhs, auto rhs) {
+                return lhs->get_function()().d() < rhs->get_function()().d();
+            });
+            return (*it)->get_function()();
+        };
+    case agg_op::avg:
+        return [refs{std::move(refs)}]() {
+            auto type = refs.front()->get_function()().type();
+            auto sum = absl::c_accumulate(
+              refs, ssmi::metric_value{0, type}, [](auto lhs, auto rhs) {
+                  return lhs + rhs->get_function()();
+              });
+            return ssmi::metric_value{
+              sum.d() / static_cast<double>(refs.size()), type};
+        };
+    };
+}
+
+ss::metrics::impl::value_map get_filtered(const filter::config& cfg) {
+    // The structure looks like this:
+    // ssmi::value_map exemplar{
+    //   {"test_fetch_metrics_counter",
+    //    ssmi::metric_family{
+    //      ssmi::metric_instances{
+    //        {ssmi::labels_type{
+    //           {"aggregate", "0"}, {"leave", "0"}, {"shard", "0"}},
+    //         ssmi::register_ref{}},
+    //        {ssmi::labels_type{
+    //           {"aggregate", "1"}, {"leave", "0"}, {"shard", "0"}},
+    //         ssmi::register_ref{}}},
+    //      ssmi::metric_family_info{
+    //        ssmi::data_type::COUNTER,
+    //        ssm::metric_type_def{""},
+    //        ssm::description{"Test a counter"},
+    //        ss::sstring{"test_fetch_metrics_counter"}}}}};
+
+    // Collect metric families
+    ssmi::value_map result;
+    for (const auto& [name, family] : ssmi::get_value_map()) {
+        auto m_it = absl::c_find_if(
+          cfg.allow, [&n = name](const auto& m) { return m.name == n; });
+
+        // Ignore metrics that are not allowed.
+        if (m_it == cfg.allow.end()) {
+            continue;
+        }
+
+        const auto& metric_cfg{*m_it};
+
+        // Collect aggregated instances
+        metric_multi_instance new_instances;
+        for (const auto& [base_labels, base_metric] : family) {
```

Comment:
There is an efficiency issue here for large partition counts: we'd be iterating over ~10E5 partition labels and summing them all up, plus the allocation overhead of building these intermediate vectors of metrics.

Reply:
This is done only at (re)configuration time, and can be made async if required.